src/client/Client.cc

```diff
@@ -5571,7 +5576,7 @@ void Client::handle_cap_grant(MetaSession *session, Inode *in, Cap *cap, const M
 
   // may drop inode's last ref
   if (deleted_inode)
-    _try_to_trim_inode(in, true);
+    _try_to_trim_inode(in, true, in->nlink == 0);
```

Comment:
I'm not sure this condition works to preserve prior behavior — if there are hard links, we might have dropped *our* last reference to the inode without it ending up at nlink==0.

Reply:
Before only when `in->nlink == 0` will the `deleted_inode` be set. The code:

 ```
5455     if (in->nlink == 0)
5456       deleted_inode = true; 
```

And currently I added another case when `CEPH_STAT_CAP_INODE_ALL` caps are being revoked will the `deleted_inode` to be set too.

Yeah, I will push another version to make the code more readable.